### ISSUE TITLE
BlockingIOError: [Errno 35] write could not complete without blocking

### ISSUE THREAD

**Issue by hipertracker** [piccolo-orm/piccolo#277]
From `piccolo shell run` the commands 

```Book.select().limit(4).run_sync()``` 

and 

```Book.select().offset(1).limit(4).run_sync()``` 

work fine, but... 

```Book.select().limit(5).run_sync()``` 

displays the result and then raise BlockingIOError exception:

```
~/.venv/lib/python3.9/site-packages/IPython/core/displayhook.py in write_format_data(self, format_dict, md_dict)
    189
    190         try:
--> 191             print(result_repr)
    192         except UnicodeEncodeError:
    193             # If a character is not supported by the terminal encoding replace

BlockingIOError: [Errno 35] write could not complete without blocking---------------------------------------------------------------------------
BlockingIOError                           Traceback (most recent call last)
```

I don't know what's going on.

**Comment by dantownsend** [comment]
Strange - do you get the same problem if you use async instead?

```python
await Book.select().limit(5).run()
```

You should be able to do this inside iPython, as it supports top level await.

I had a dig around, and `BlockingIOError` is raised when a blocking operation in used in a non-blocking context.

From a Piccolo perspective, we sometimes use the `print` statement, which is blocking, so could potentially be the problem if it's printing out lots of data.

**Comment by dantownsend** [comment]
Another idea, maybe `run_sync` is being used within an `async` function?

**Comment by hipertracker** [comment]
```await Book.select().limit(5).run()```  raise the same error ```BlockingIOError: [Errno 35] write could not complete without blocking``` 

This is a small table with only 95 records. I can read up to 4 records, when I read 5 or more I get that error. This is very strange.

**Comment by hipertracker** [comment]
This is my model, nothing complicated:
```python
class Book(Table, tablename="books"):
    name = Varchar(length=32, unique=True)
    label = JSONB(default=dict())
    sort = Integer(default=0, index=True)
    available = Boolean(default=True, index=True)
    tags = Array(base_column=Varchar(length=32), default=[])
```

**Comment by dantownsend** [comment]
On the traceback it said something about a unicode character - there isn't a strange character in your database which is causing issues?

Do you still get the same problem in a standard Python shell, or just with iPython?

**Comment by dantownsend** [comment]
I wrote a script to test it, and can't replicate the issue:

```python
import asyncio

from faker import Faker
from piccolo.table import Table
from piccolo.columns.column_types import Varchar, JSONB, Integer, Boolean, Array
from piccolo.engine.postgres import PostgresEngine


DB = PostgresEngine(
    config={
        "database": "piccolo_experiments",
        "user": "postgres",
        "password": "",
        "host": "localhost",
        "port": 5432,
    }
)


class Book(Table, tablename="books", db=DB):
    name = Varchar(length=32, unique=True)
    label = JSONB(default=dict())
    sort = Integer(default=0, index=True)
    available = Boolean(default=True, index=True)
    tags = Array(base_column=Varchar(length=32), default=[])


async def main():
    await Book.alter().drop_table(if_exists=True).run()
    await Book.create_table().run()

    fake = Faker()
    Faker.seed(0)

    await Book.insert(
        *[
            Book(
                name=fake.sentence(5),
                label={'code': fake.pyint()},
                sort=fake.pyint(),
                available=fake.boolean(),
                tags=[fake.word(), fake.word(), fake.word()]
            ) or _ in range(1000)
        ]
    ).run()

    print(await Book.select().limit(5).run())


if __name__ == '__main__':
    asyncio.run(main())
```

So maybe it's a quirk of iPython, or a data issue?

**Comment by hipertracker** [comment]
It's a problem with iPython.  I can run the code from a file or with a standard Python console, but not with iPython.

**Comment by hipertracker** [comment]
BTW, it works perfectly in Jupyter Notebook.

**Comment by dantownsend** [comment]
@hipertracker I encountered the same issue randomly when running a query in iPython.

I found this related issue:

https://github.com/ipython/ipython/issues/12563

Seems to be some weird issue when uvloop is used with iPython on Mac OS.

Uninstalling uvloop fixed it for me.

**Comment by hipertracker** [comment]
Piccolo is installing uvloop by default. How to disable it? I use only this in requirements.txt and still uvloop is installed
```
piccolo_admin
piccolo[postgres]
piccolo[orjson]
```

**Comment by dantownsend** [comment]
Hmm, strange.

I just tried `pip install piccolo_admin piccolo[postgres,orjson]` in a Docker container, and this is what I got.

<img width="284" alt="Screenshot 2021-09-30 at 08 11 27" src="https://user-images.githubusercontent.com/350976/135404675-799a27cf-9411-48fd-82dd-08b69b69842c.png">

If you use `pip install piccolo[all]`, it will definitely install uvloop, but shouldn't otherwise. Unless it's being installed by some other dependency.
